### PR TITLE
add RAFT

### PR DESCRIPTION
**Pull request by DearAJ**
Set "adv_estimator": "raft" in config.algorithm to enable support for the RAFT algorithm.

### CODE REVIEW

File: agentlightning/verl/trainer.py [hunk 1]
```diff
@@ -388,6 +392,251 @@ def _train_step(self, batch_dict: dict) -> dict:
...
+            n_remained_transition = n_transition // mini_batch_size * mini_batch_size
+            positive_batch = positive_batch[list(range(n_remained_transition))]
+            metrics["raft/n_triplets_dropped_remainder"] = n_transition - n_remained_transition
+            
+            # Balance batch if enabled
+            if self.config.trainer.balance_batch:
+                self._balance_batch(positive_batch, metrics=metrics)
+            
+            # Pad batch for distributed training
+            positive_batch, pad_size = pad_dataproto_to_divisor(positive_batch, self.actor_rollout_wg.world_size)
```
Comment (XufangLuo) [inline agentlightning/verl/trainer.py]: Why three paddings (here and L514, L577) are needed?

File: agentlightning/verl/trainer.py [hunk 2]
```diff
@@ -388,6 +392,251 @@ def _train_step(self, batch_dict: dict) -> dict:
...
+                # This makes the PPO loss equivalent to standard cross-entropy
+                raft_batch.batch["advantages"] = torch.ones(
+                    (len(raft_batch), max_response_length),
+                    device=input_ids.device,
+                    dtype=torch.float32
+                )
+                raft_batch.batch["returns"] = raft_batch.batch["advantages"].clone()
+                
+                # Store labels in batch for potential use
+                raft_batch.batch["labels"] = labels
```
Comment (XufangLuo) [inline agentlightning/verl/trainer.py]: even labels are specified here. will these labels be used by the following update_actor method? or will they be used as expected?

File: agentlightning/verl/trainer.py [hunk 3]
```diff
@@ -388,6 +392,251 @@ def _train_step(self, batch_dict: dict) -> dict:
...
+                
+                # Disable clipping: set both ratios to 1.0 (no clipping in pure SFT)
+                self.config.actor_rollout_ref.actor["clip_ratio_low"] = 1.0
+                self.config.actor_rollout_ref.actor["clip_ratio_high"] = 1.0
+                
+                try:
+                    # Update actor with pure SFT loss
+                    # With advantages=1.0 and clip_ratio=1.0, this becomes standard cross-entropy
+                    # This mimics SFTTrainer.compute_loss() behavior
+                    actor_output = self.actor_rollout_wg.update_actor(raft_batch)
```
Comment (XufangLuo) [inline agentlightning/verl/trainer.py:564]: the called update_actor function is still the RL one, not the one in "SFTTrainer"? not sure if verl has SFTTrainer...

File: agentlightning/verl/trainer.py [hunk 4]
```diff
@@ -388,6 +392,251 @@ def _train_step(self, batch_dict: dict) -> dict:
...
+                
+                # Pad again for distributed training before update_actor
+                raft_batch, pad_size_actor = pad_dataproto_to_divisor(raft_batch, self.actor_rollout_wg.world_size)
+                
+                # Temporarily disable PPO clipping for pure SFT (like SFTTrainer)
+                # Set clip_ratio to 1.0 effectively disables clipping
+                original_clip_low = self.config.actor_rollout_ref.actor.get("clip_ratio_low", 0.2)
+                original_clip_high = self.config.actor_rollout_ref.actor.get("clip_ratio_high", 0.3)
+                
+                # Disable clipping: set both ratios to 1.0 (no clipping in pure SFT)
```
Comment (XufangLuo) [inline agentlightning/verl/trainer.py]: why clipping can be disabled by setting them to 1.0? This might not be related to RAFT.

File: agentlightning/verl/trainer.py (hunk 3 above)
Comment (DearAJ) [reply agentlightning/verl/trainer.py:564]: Verl does not have a SFTrainer. SFTrainer inherits from transformers.Trainer and requires the complete HuggingFace Trainer infrastructure, whereas VERL uses Ray distributed training and a custom worker group. Directly adopting SFTrainer would disrupt VERL's existing architecture. Additionally, SFTrainer and VERL use different data formats.

File: agentlightning/verl/trainer.py (hunk 4 above)
Comment (DearAJ) [reply agentlightning/verl/trainer.py]: Here is the derivation of SFT from PPO：

The standard PPO objective (with clipping) is defined as:
L = - E(min( rA, clip( r， 1-ϵ，1+ϵ )A ))

To adapt the PPO framework for Supervised Fine-Tuning (SFT) mode, we set the following neutral conditions:
* Set advantages:  A=1
* Disable clipping: ϵ = 1.0 
   clip(r, 1-1, 1+0) = clip(r, 0, 2) ≈ r  (r = exp(log_prob - old_log_prob), which is usually ranges in [0, 2])

When A=1 and clipping is disabled, the loss simplifies to:
L = - E(min( r, clip( r， 1-ϵ，1+ϵ ))) = -E(r)
logr = logπ(a|s)-logπold(a|s)

Therefore:
L = - E( exp(logπ(a|s)-logπold(a|s)))

In SFT, we directly optimize the current policy without relying on importance sampling. When the old policy is equal to (or very close to) the current policy, the PPO objective is replaced with the Negative Log-Likelihood  loss, which is what we want to minimize:
L = - E( logπ(a|s))
For language models, this is the standard Cross-Entropy Loss.

Summary, by:
* Setting A=1.0$to remove advantage weighting.
* Disabling clipping to remove the PPO clipping mechanism.
The final loss effectively degenerates (or is replaced by) the standard Cross-Entropy Loss (Negative Log-Likelihood), which is the SFT loss.
Thus, the PPO framework, under these specific conditions, becomes equivalent to standard supervised learning (SFT).

File: agentlightning/verl/trainer.py (hunk 3 above)
Comment (XufangLuo) [reply agentlightning/verl/trainer.py:564]: Make sense. Then why SFT can be implemented in this way? Will update_actor actually compute SFT loss as we expected?

File: agentlightning/verl/trainer.py (hunk 1 above)
Comment (DearAJ) [reply agentlightning/verl/trainer.py]: > Why three paddings (here and L514, L577) are needed?

The first padding operation was a legacy from the previous code and has now been removed. The latter two padding operations are necessary:

compute_log_prob requires distributed training, and the batch size must be divisible by the world_size.

update_actor requires distributed training, and the batch size must be divisible by the world_size.

File: agentlightning/verl/trainer.py (hunk 2 above)
Comment (DearAJ) [reply agentlightning/verl/trainer.py]: > even labels are specified here. will these labels be used by the following update_actor method? or will they be used as expected?

These labels will not be used, but the update_actor method requires these fields. Therefore, I have retained the original GRPO processing logic to ensure the data format meets the requirements.

File: agentlightning/verl/trainer.py (hunk 3 above)
Comment (DearAJ) [reply agentlightning/verl/trainer.py:564]: > Make sense. Then why SFT can be implemented in this way? Will update_actor actually compute SFT loss as we expected?

Here is the derivation of SFT from PPO：

The standard PPO objective (with clipping) is defined as:
L = - E(min( rA, clip( r， 1-ϵ，1+ϵ )A ))

To adapt the PPO framework for Supervised Fine-Tuning (SFT) mode, we set the following neutral conditions:

Set advantages: A=1
Disable clipping: ϵ = 1.0
clip(r, 1-1, 1+0) = clip(r, 0, 2) ≈ r (r = exp(log_prob - old_log_prob), which is usually ranges in [0, 2])
When A=1 and clipping is disabled, the loss simplifies to:
L = - E(min( r, clip( r， 1-ϵ，1+ϵ ))) = -E(r)
logr = logπ(a|s)-logπold(a|s)

Therefore:
L = - E( exp(logπ(a|s)-logπold(a|s)))

In SFT, we directly optimize the current policy without relying on importance sampling. When the old policy is equal to (or very close to) the current policy, the PPO objective is replaced with the Negative Log-Likelihood loss, which is what we want to minimize:
L = - E( logπ(a|s))
For language models, this is the standard Cross-Entropy Loss.

Summary, by:

Setting A=1.0$to remove advantage weighting.
Disabling clipping to remove the PPO clipping mechanism.
The final loss effectively degenerates (or is replaced by) the standard Cross-Entropy Loss (Negative Log-Likelihood), which is the SFT loss.
Thus, the PPO framework, under these specific conditions, becomes equivalent to standard supervised learning (SFT).

File: agentlightning/verl/trainer.py (hunk 4 above)
Comment (XufangLuo) [reply agentlightning/verl/trainer.py]: Here are some things I am not very sure.

1. clipping is applied to importance sampling ratio, which is a ratio between prob and old_prob. So the the numerical range of importance sampling ratio is from 0 to positive infinity. Setting high and low to 1 leads to clip(r, 0, 2), which may not disabled clipping.
2. If you want to use A to implement RAFT and SFT loss, what are the "labels" used for?

Comment (DearAJ) [reply agentlightning/verl/trainer.py]: Thank you for pointing out the issues. I have pushed new modifications.

1. To simplify the formula, I previously assumed π ≈ π_old and set clip_ratio_low and clip_ratio_high to the same value. In fact, a more rigorous way to disable clipping would be to set clip_ratio_high to an extremely large number.

2. Initially, I intended to calculate the loss following the implementation of SFTrainer. However, I later changed the calculation method, resulting in variables from both approaches coexisting. The "labels" were redundant code and have now been removed.
